astropy/units/tests/test_quantity_ufuncs.py

```diff
@@ -1294,6 +1295,91 @@ def test_two_argument_ufunc_outer(self):
         assert np.all(s13_greater_outer == check13_greater_outer)
 
 
+@dataclasses.dataclass
+class DuckQuantity1:
+    data: u.Quantity
+
+
+@dataclasses.dataclass
+class DuckQuantity2(DuckQuantity1):
+    @property
+    def unit(self) -> u.UnitBase:
+        return self.data.unit
+
+
+@dataclasses.dataclass(eq=False)
+class DuckQuantity3(DuckQuantity2):
+    def __array_ufunc__(self, function, method, *inputs, **kwargs):
+
+        inputs = [inp.data if isinstance(inp, type(self)) else inp for inp in inputs]
+
+        kwargs_copy = {}
+        for k in kwargs:
+            kwarg = kwargs[k]
+            if isinstance(kwarg, type(self)):
+                kwargs_copy[k] = kwarg.data
+            elif isinstance(kwarg, (list, tuple)):
+                kwargs_copy[k] = type(kwarg)(
+                    item.data if isinstance(item, type(self)) else item
+                    for item in kwarg
+                )
+            else:
+                kwargs_copy[k] = kwarg
+        kwargs = kwargs_copy
+
+        for inp in inputs:
+            if isinstance(inp, np.ndarray):
+                result = inp.__array_ufunc__(function, method, *inputs, **kwargs)
+                if result is not NotImplemented:
+                    return type(self)(result)
+
+        return NotImplemented
+
+
+@pytest.mark.parametrize(
+    "duck_quantity",
+    (DuckQuantity1(1 * u.mm), DuckQuantity2(1 * u.mm), DuckQuantity3(1 * u.mm)),
+)
+@pytest.mark.parametrize("out", (None, "empty"))
+class TestUfuncReturnsNotImplemented:
+    @pytest.mark.parametrize("ufunc", (np.negative, np.abs))
```

Comment:
A bit of a nitpick, but I think the cases for the real duck, `DuckQuantity3`, are sufficiently different that they would be better done as separate methods (so one avoids the `if/else`). I'm fine with not testing `out` for `DuckQuantity[12]`.

Reply:
Ok, I've pushed an update that has separate tests for `DuckQuantity[12]` and `DuckQuantity3`, let me know what you think.